Code/GraphMol/SmilesParse/CXSmilesOps.cpp

```diff
@@ -356,3 +357,148 @@ void parseCXExtensions(RDKit::RWMol &mol, const std::string &extText,
+      res += lbl;
+    }
+  }
+  return res;
+}
+std::string get_radical_block(const ROMol &mol,
+                              const std::vector<unsigned int> &atomOrder) {
+  std::string res = "";
+  std::map<unsigned int, std::vector<unsigned int>> rads;
+  for (unsigned int i = 0; i < mol.getNumAtoms(); ++i) {
+    auto idx = atomOrder[i];
+    auto nrad = mol.getAtomWithIdx(idx)->getNumRadicalElectrons();
+    if (nrad) {
+      rads[nrad].push_back(i);
+    }
+  }
+  if (rads.size()) {
+    for (const auto pr : rads) {
+      switch (pr.first) {
+        case 1:
+          res += "^1:";
+          break;
+        case 2:
+          res += "^2:";
+          break;
+        case 3:
+          res += "^5:";
+          break;
+        default:
+          BOOST_LOG(rdWarningLog) << "unsupported number of radical electrons "
+                                  << pr.first << std::endl;
+      }
+      for (auto aidx : pr.second) {
+        res += boost::str(boost::format("%d,") % aidx);
+      }
+    }
+  }
+  return res;
+}
+std::string get_coords_block(const ROMol &mol,
+                             const std::vector<unsigned int> &atomOrder) {
+  std::string res = "";
+  const auto &conf = mol.getConformer();
+  bool first = true;
+  for (auto idx : atomOrder) {
+    const auto &pt = conf.getAtomPos(idx);
+    if (!first) {
+      res += ";";
+    } else {
+      first = false;
+    }
+    res += boost::str(boost::format("%g,%g,") % pt.x % pt.y);
+    if (conf.is3D()) {
+      auto zc = boost::str(boost::format("%g") % pt.z);
+      if (zc != "0") res += zc;
+    }
+  }
+  return res;
+}
+
+std::string get_atom_props_block(const ROMol &mol,
+                                 const std::vector<unsigned int> &atomOrder) {
+  std::vector<std::string> skip = {common_properties::atomLabel,
+                                   common_properties::molFileValue};
+  std::string res = "";
+  unsigned int which = 0;
+  for (auto idx : atomOrder) {
+    const auto atom = mol.getAtomWithIdx(idx);
+    bool includePrivate = false, includeComputed = false;
+    for (const auto &pn : atom->getPropList(includePrivate, includeComputed)) {
+      if (std::find(skip.begin(), skip.end(), pn) == skip.end()) {
+        if (res.size() == 0) res += "atomProp";
+        res += boost::str(boost::format(":%d.%s.%s") % which % pn %
+                          (atom->getProp<std::string>(pn)));
+      }
+    }
+    ++which;
+  }
+  return res;
+}
+}  // namespace
+std::string getCXExtensions(const ROMol &mol) {
+  std::string res = "|";
+  // we will need atom ordering. Get that now:
+  std::vector<unsigned int> atomOrder = mol.getProp<std::vector<unsigned int>>(
+      common_properties::_smilesAtomOutputOrder);
+  bool needLabels = false;
+  bool needValues = false;
+  for (const auto at : mol.atoms()) {
+    if (at->hasProp(common_properties::atomLabel)) needLabels = true;
+    if (at->hasProp(common_properties::molFileValue)) needValues = true;
```

Comment:
maybe add
```
if(needsLabels&&needsValues) break;
```

Reply:
If/when I complete the support for CXSMILES/CXSMARTS, there may be additional code in this loop, making that test longer and longer. Given that the loop isn't expensive, it seems like an unnecessary optimization.